ql/cashflows/overnightindexedcoupon.cpp

```diff
@@ -40,25 +41,34 @@ namespace QuantLib {
                 coupon_ = dynamic_cast<const OvernightIndexedCoupon*>(&coupon);
                 QL_ENSURE(coupon_, "wrong coupon type");
             }
-            Rate swapletRate() const override {
+            Rate swapletRate(const Date& date) const override {
+
+                const Date today = Settings::instance().evaluationDate();
 
-                ext::shared_ptr<OvernightIndex> index =
+                const ext::shared_ptr<OvernightIndex> index =
                     ext::dynamic_pointer_cast<OvernightIndex>(coupon_->index());
 
                 const vector<Date>& fixingDates = coupon_->fixingDates();
+                const vector<Date>& valueDates = coupon_->valueDates();
                 const vector<Time>& dt = coupon_->dt();
 
-                Size n = dt.size(),
-                     i = 0;
+                Size i = 0;
+                auto endDateIterator = valueDates.end();
+                
+                if (date < *endDateIterator) {
+                    endDateIterator = std::lower_bound(valueDates.begin(), valueDates.end(), date);
+                }
 
+                const auto endDate = *endDateIterator;
+                const size_t n = endDateIterator - valueDates.begin();
                 Real compoundFactor = 1.0;
 
+                const auto minDate = std::min(today, date);
+
                 // already fixed part
-                Date today = Settings::instance().evaluationDate();
-                while (i<n && fixingDates[i]<today) {
+                while (i < n && fixingDates[i] < minDate) {
```

Comment:
You don't need `minDate`.  If `date` is before `today`, `i` will equal `n` and exit the loop before reaching today.
The same would happen in the original code is the end date of the coupon was before today.